homeassistant/components/yeelight/__init__.py

```diff
@@ -139,73 +152,252 @@
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+
+    if unload_ok:
+        if entry.data[CONF_DISCOVERY]:
+            data = hass.data[DOMAIN][DATA_CONFIG_ENTRIES].pop(entry.entry_id)
+            scanner = data[DATA_SCANNER]
+            await scanner.stop_scan()
+            for ipaddr in scanner.seen:
+                hass.data[DOMAIN][DATA_DEVICES].pop(ipaddr)
+            data[DATA_UNSUB_UPDATE_LISTENER]()
+        else:
+            hass.data[DOMAIN][DATA_DEVICES].pop(entry.data[CONF_IP_ADDRESS])
+            data = hass.data[DOMAIN][DATA_CONFIG_ENTRIES].pop(entry.entry_id)
+            data[DATA_UNSUB_UPDATE_LISTENER]()
+
+    return unload_ok
+
+
+async def _async_setup_device(
+    hass: HomeAssistant,
+    ipaddr: str,
+    config: dict,
+    discovery_config_entry: Optional[ConfigEntry] = None,
+) -> None:
+    # Set up device
+    bulb = Bulb(ipaddr, model=config.get(CONF_MODEL) or None)
+    capabilities = await hass.async_add_executor_job(bulb.get_capabilities)
+    if capabilities is None:  # timeout
+        _LOGGER.error("Failed to get capabilities from %s", ipaddr)
+        raise ConfigEntryNotReady
+    device = YeelightDevice(hass, ipaddr, config, bulb)
+    await hass.async_add_executor_job(device.update)
+    hass.data[DOMAIN][DATA_DEVICES][ipaddr] = device
+
+    # Trigger platform setup
+    if discovery_config_entry is not None:
+        data = hass.data[DOMAIN][DATA_CONFIG_ENTRIES][discovery_config_entry.entry_id]
+        hass.async_create_task(data[DATA_SETUP_BINARY_SENSOR](ipaddr))
+        hass.async_create_task(data[DATA_SETUP_LIGHT](ipaddr))
+
+
+async def _async_update_listener(hass: HomeAssistant, entry: ConfigEntry):
+    """Handle options update."""
+    await hass.config_entries.async_reload(entry.entry_id)
+
+
+class YeelightScanner:
+    """Scan for Yeelight devices."""
+
+    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry):
+        """Initialize class."""
+        self._hass = hass
+        self._config_entry = config_entry
+        self._remove_listender = None
+        self.seen = set()
+
+    async def _async_scan(self, _):
+        _LOGGER.debug("Yeelight scanning")
+        # Run 3 times as packets can get lost
+        for _ in range(3):
+            devices = await self._hass.async_add_executor_job(discover_bulbs)
+            for device in devices:
+                unique_id = device["capabilities"]["id"]
+                ipaddr = device["ip"]
+                if ipaddr in self.seen:
+                    continue
+                self.seen.add(ipaddr)
+                _LOGGER.debug("Yeelight discoevered at %s", ipaddr)
+
+                # Initialize default configuration
+                if unique_id not in self._config_entry.options:
+                    # Temporarily disable update listener to avoid reload
```

Comment:
Is each device one config entry? Is that necessary? It's not possible to always use the discovery method and have an options flow to select between discovered devices?

Reply:
This code here is to add default configuration to options for each newly discovered device, so that I don't have to handle the special cases in entry setup and options flow. Turns out the code here ends up quite complicated and requires enabling / disabling the listeners and reloading the entry several times. What do you suggest? Maybe it is better to just handle default configuration in entry setup?

Reply:
We normally set default options in config entry setup.